### PR TITLE
Temporary Queues and Topics

### PR DESCRIPTION
**Pull request by adamretter**
I would like the facility to work with JMS temporary queues. This is so that I can then implement something like the [EAI Request Response pattern](https://www.enterpriseintegrationpatterns.com/RequestReplyJmsExample.html) atop jms4s for our own application.

Would this be something that would be welcome as a Pull Request to this project?

I thought this might be a small addition to jms4s to add this functionality... Unfortunately, I needed to add some additional API so that the Temporary Queue/Topic can be created just once and then reused for the Consumer and Producer(s).

### CODE REVIEW

**Comment by faustin0** [comment]
Hi @adamretter, thanks for your contributions! I will take a look at the changes and run the tests trying to understand what's going on. 
From a first look a the ibmMQ error it seems that some permissions are not set in the "scripts/definitions.mqsc"

File: core/src/main/scala/jms4s/config/config.scala [hunk 1]
```diff
@@ -23,7 +23,21 @@ package jms4s.config
...
 import cats.Order
 
-sealed trait DestinationName        extends Product with Serializable
+import java.util.UUID
+
+sealed trait Destination
+
+sealed trait TemporaryDestination extends Destination
+
+case class TemporaryQueue(value: UUID = java.util.UUID.randomUUID()) extends TemporaryDestination {
```
Comment (faustin0) [inline core/src/main/scala/jms4s/config/config.scala:32]: Creating an UUID is an impure and blocking operation, i will not make it a default value.
It should be executed inside a "blocking" block or, better, use https://typelevel.org/cats-effect/api/3.x/cats/effect/std/UUIDGen$.html.
This case-class should be a simple container of the name, the generation for the UUID should be moved elsewhere and executed in an appropriate context (an F[_] that has this capability).
If we want to maintain a default/utility method to create a `TemporaryQueue` with an UUID we can create a method(that works in `F[_]: UUIDGen` ) in the companion object .

Comment (adamretter) [reply core/src/main/scala/jms4s/config/config.scala]: @faustin0 Good point! I have pushed a commit to try and improve this. Please let me know your thoughts...

**Comment by adamretter** [comment]
@faustin0 All tests are passing and I hope this is now in good shape for your review... please let me know what you think?

File: core/src/main/scala/jms4s/jms/JmsContext.scala [hunk 2]
```diff
@@ -88,12 +109,22 @@ class JmsContext[F[_]: Async: Logger](private val context: JMSContext) {
...
     Sync[F].blocking(context.createTopic(topicName.value)).map(new JmsTopic(_))
 
-  def createDestination(destination: DestinationName): F[JmsDestination] = destination match {
-    case q: QueueName => createQueue(q).widen[JmsDestination]
-    case t: TopicName => createTopic(t).widen[JmsDestination]
+  private def createTemporaryTopic(): F[JmsTopic] =
+    Sync[F].blocking(context.createTemporaryTopic()).map(new JmsTopic(_))
+
+  override def createDestination(destination: Destination): F[JmsDestination] = destination match {
+    case TemporaryQueue(_) => createTemporaryQueue().widen[JmsDestination]
```
Comment (faustin0) [inline core/src/main/scala/jms4s/jms/JmsContext.scala:124]: hi @adamretter i'm giving a closer look at the PR 🙂 
it seems that the UUID value inside `TemporaryQueue` and `TemporaryTopic` is never used outside of the .`toString`, and i can't really see were this UUID(and the toString) can be usefull in a config class that is used to create the actual JMS TemporaryQueue.
like here, it's ignored.
I'm missing something or we can remove it and simplify the api?

Comment (adamretter) [reply core/src/main/scala/jms4s/jms/JmsContext.scala:124]: @faustin0 It was previously used by a Logger to help understand the message queues that are created. It is no longer used and can now be removed. Apart from that do you have any other comments for me, or are you happy with the changes?

**Comment by faustin0** [comment]
@adamretter i'm having troubles creating a consumer (with a concurrency level >1 ) for a temp Q  that use the `JmsDestination`, it fails with an MQ error.
I'm not very familiar with the use of temp Q, are multiple consumers allowed for a temp Q ? 
if yes, could you add a test where a consumer(for a temp Q) is created with a concurrency > 1 ? Like in the JmsClientSpec suite (a more high level test than using context directly)

**Comment by faustin0** [comment]
hi @adamretter, as you can see, the build is failing for the new test (with the same error i was getting in my local testing):
`com.ibm.mq.MQException: JMSCMQ0001: IBM MQ call failed with compcode '2' ('MQCC_FAILED') reason '2042' `('MQRC_OBJECT_IN_USE').
Isn'it failing on your local testing?

**Comment by adamretter** [comment]
@faustin0 Sorry, I only got as far as adding the additional tests, I have not yet looking into the issue that causes the test to fail. I also still see locally this error:

```
[info] - publish 50 messages to a temporary queue and then consume them concurrently with local transactions *** FAILED ***
[info]   com.ibm.msg.client.jms.DetailedIllegalStateRuntimeException: JMSWMQ2008: Failed to open MQ queue 'AMQ.63BFFFB125461803'.
[info]   at com.ibm.msg.client.jms.DetailedIllegalStateException.getUnchecked(DetailedIllegalStateException.java:274)
[info]   at com.ibm.msg.client.jms.internal.JmsErrorUtils.convertJMSException(JmsErrorUtils.java:173)
[info]   at com.ibm.msg.client.jms.internal.JmsContextImpl.createConsumer(JmsContextImpl.java:420)
[info]   at jms4s.jms.JmsContext.$anonfun$createJmsConsumer$3(JmsContext.scala:95)
[info]   at blocking @ jms4s.ibmmq.ibmMQ$.$anonfun$makeJmsClient$2(ibmMQ.scala:56)
[info]   at delay @ org.typelevel.log4cats.slf4j.internal.Slf4jLoggerInternal$Slf4jLogger.$anonfun$info$4(Slf4jLoggerInternal.scala:91)
[info]   at delay @ org.typelevel.log4cats.slf4j.internal.Slf4jLoggerInternal$Slf4jLogger.isInfoEnabled(Slf4jLoggerInternal.scala:66)
[info]   at ifM$extension @ org.typelevel.log4cats.slf4j.internal.Slf4jLoggerInternal$Slf4jLogger.info(Slf4jLoggerInternal.scala:91)
[info]   at *> @ jms4s.ibmmq.ibmMQ$.$anonfun$makeJmsClient$7(ibmMQ.scala:73)
[info]   at make @ jms4s.ibmmq.ibmMQ$.makeJmsClient(ibmMQ.scala:71)
[info]   ...
[info]   Cause: com.ibm.mq.MQException: JMSCMQ0001: IBM MQ call failed with compcode '2' ('MQCC_FAILED') reason '2042' ('MQRC_OBJECT_IN_USE').
[info]   at com.ibm.msg.client.wmq.common.internal.Reason.createException(Reason.java:203)
[info]   at com.ibm.msg.client.wmq.internal.WMQMessageConsumer.checkJmqiCallSuccess(WMQMessageConsumer.java:222)
[info]   at com.ibm.msg.client.wmq.internal.WMQMessageConsumer.checkJmqiCallSuccess(WMQMessageConsumer.java:156)
[info]   at com.ibm.msg.client.wmq.internal.WMQConsumerShadow.initialize(WMQConsumerShadow.java:1173)
[info]   at com.ibm.msg.client.wmq.internal.WMQSyncConsumerShadow.initialize(WMQSyncConsumerShadow.java:128)
[info]   at com.ibm.msg.client.wmq.internal.WMQMessageConsumer.<init>(WMQMessageConsumer.java:486)
[info]   at com.ibm.msg.client.wmq.internal.WMQSession.createConsumer(WMQSession.java:929)
[info]   at com.ibm.msg.client.jms.internal.JmsSessionImpl.createConsumer(JmsSessionImpl.java:1012)
[info]   at com.ibm.msg.client.jms.internal.JmsSessionImpl.createConsumer(JmsSessionImpl.java:1097)
[info]   at com.ibm.msg.client.jms.internal.JmsContextImpl$SessionWrapper.createConsumer(JmsContextImpl.java:1989)
[info]   ...
```

**Comment by faustin0** [comment]
superseded by https://github.com/fp-in-bo/jms4s/pull/377

**Pull request closed without merging**